### PR TITLE
fix: anyOf produces invalid json due to extra comma

### PR DESCRIPTION
**Pull request by samzilverberg**
PR to address https://github.com/fastify/fast-json-stringify/issues/642

It simple removes the extra comma that was introduced since https://github.com/fastify/fast-json-stringify/pull/630 .

I spent around an hour trying a better approach where the comma is not added in the first place.
but the code is hard to follow and I don't have enough time to get to the bottom of it.

added test case from the issue (was red before fix, green after)

### CODE REVIEW

**Comment by samzilverberg** [comment]
I spent a little more time on figuring out a better way that would simple not add the commas if not needed.
It didn't work out.
The problem was with this code:
```
      code += `
        ${index === 0 ? 'if' : 'else if'}(validator.validate("${schemaRef}", ${input}))
          ${nestedResult}
      `
```

for the "anyOf" kind of block that contains "required", the "nestedResult" is just a block of code that either produces an err or adds nothing to the json.
but when the anyOf contains other schema, the nestedResult can also add something to the json.

this means the comma is dependand on what the "nestedResult" does.

the nestedResult returns something like `json += blabla` (or json += "") , and not just the value, so I can't put a condition on it to add the comma :(

a major refactor would be needed to make this work....

posting all of this to convince you this fix is the only way I see forward, unless you are interested in a major refactor, or maybe you can think of another fix...

**Comment by Uzlopak** [comment]
Yeah, I see that you also dislike the solution. Maybe you try to fix it from the wrong angle?

If I remove this line

https://github.com/fastify/fast-json-stringify/blob/4073e5b6c9b2f4c6b4a3cea1f0f1d6f0ba89e769/index.js#L876

Then your unit test works without your changes.

But it will fail 'object with anyOf nested inside allOf' in allof.test.js, because then a comma is missing. So maybe, it has to be fixed for that case and not for the case you implemented. So maybe we can detect better, when we need to add a comma for this allof case instead of generally setting the comma and being forced to remove the comma in an edge case.

**Comment by samzilverberg** [comment]
yes,  I ran both tests and ran some experiments with adding commas based on different conditions (and not doing the substring i did in this pr).
It always boiled down to this if condition i posted and whether the "nestedResult" added something to the json or not (which is not detectable atm :( )

I don't have a better idea atm.

**Comment by sloops77** [comment]
I've played around with this locally and I may have a solution. All tests pass including the new one that is  a part of this PR

```
...
  if (schema.type === 'object') {
      context.wrapObjects = false
      const funcName = buildObject(context, location)
      code += `
        json += '{'
        json += ${funcName}(${input})       
      `
    }

    const type = schema.anyOf ? 'anyOf' : 'oneOf'
    const anyOfLocation = location.getPropertyLocation(type)

    if (schema.type === 'object' && location.schema[type].length > 0) {
      code += `
        json += ','
      `
    }
    ...
```

**Comment by mcollina** [comment]
@sloops77 could you send a fresh PR?

**Comment by samzilverberg** [comment]
closing in favor of more proper solution in https://github.com/fastify/fast-json-stringify/pull/651
